components/feature/autofill/src/main/java/mozilla/components/feature/autofill/AbstractAutofillService.kt

```diff
@@ -60,10 +64,33 @@ abstract class AbstractAutofillService : AutofillService() {
     }
 
     override fun onSaveRequest(request: SaveRequest, callback: SaveCallback) {
-        // This callback should not get invoked since we do not indicate that we are interested in
-        // saving any data (yet). If for whatever reason it does get invoked then we pretent that
-        // we handled the request successfully. Calling onFailure() requires to pass in a message
-        // and on Android systems before Q this message may be shown in a toast.
+        var username: String? = null
+        var password: String? = null
+        var parsedStructure: ParsedStructure? = null
+        for (fillContext in request.fillContexts) {
+            parsedStructure = parseStructure(this, fillContext.structure.toRawStructure())
+            parsedStructure?.also { ps ->
+                ps.username?.let { username = it }
+                ps.password?.let { password = it }
+            }
+        }
+        parsedStructure ?: return
+        username ?: return
+        password ?: return
+
+        @OptIn(DelicateCoroutinesApi::class)
+        GlobalScope.launch(Dispatchers.IO) {
+            val lookupDomain = parsedStructure.getLookupDomain(configuration.publicSuffixList)
+            val loginEntry = Login(
+                guid = "",
+                origin = "https://$lookupDomain",
+                username = username!!,
+                password = password!!,
+                httpRealm = "https://$lookupDomain"
+            ).toEntry()
+            configuration.storage.addOrUpdate(loginEntry)
```

Comment:
cc @grigoryk: Can you review the saving piece here? Is that how we should do it?

Reply:
As it says on the lid, this will try to look up an existing login and update it, or if nothing matches, it will create a new one.

Here's how we match an existing record:
https://github.com/mozilla/application-services/blob/a5613c64dcf13fb12fdbce766784ca0250c018c7/components/logins/src/db.rs#L224-L258

and https://github.com/mozilla/application-services/blob/a5613c64dcf13fb12fdbce766784ca0250c018c7/components/logins/src/db.rs#L617-L685

... or in English: look at _all_ records for which `origin` and (in this case) `httpRealm` match, then pick a record with the same `username`, or a blank username as a fallback.

So, I guess in context of saving logins from other apps, `origin` and `httpRealm` need to be at least stable for the app, e.g. not change from one interaction to another.

And perhaps they should be unique to the app, as well? To avoid clashes with other applications... And maybe not something the application itself can control, to avoid security/spoofing issues.

... and now that I actually looked at what `getLookupDomain` does, I think we're okay? `package name of the application transformed into a domain`.

Would it make sense to also perform a digital assets link at the point of saving? E.g. verify that the app is actually linked to the `lookupDomain`... /me reads https://github.com/google/digitalassetlinks/blob/master/well-known/details.md to see if that even makes any sense..

Reply:
Yeah, it makes sense. E.g. for Twitter, digitalassets is at https://twitter.com/.well-known/assetlinks.json - and lists the two android apps (twitter & twitter lite).

Here, our lookupDomain will be `http://android.twitter.com`. In this particular case we can fallback to the root domain, which will work... Does that generalize..?

Reply:
In terms of our `DigitalAssetLinksApi`, how do we obtain `AssetDescriptor.Web` to run the verification?

Reply:
Interesting!
I'll look into it

Reply:
Hey @grigoryk 
I see that the check is performed when filling out the form inside [`FillRequestHandler.handle()`](https://github.com/VitalyVPinchuk/android-components/blob/cecb82fd82e6676ec1a49dd62a4ef9aedc2fd048/components/feature/autofill/src/main/java/mozilla/components/feature/autofill/handler/FillRequestHandler.kt#L64)

Should we perform the same check when saving credentials?

Reply:
And it looks like the following

<img src="https://user-images.githubusercontent.com/26341194/140966270-e1a06311-75ae-4cba-9626-fdf2aa76a60a.png" alt="Verification failed screenshot" width="300">